apollo-router-core/src/query_planner/mod.rs

```diff
@@ -253,65 +269,127 @@ impl PlanNode {
 
-    /// The GraphQL subquery that is used for the fetch.
-    operation: String,
-}
+    /// A fetch node.
+    #[derive(Debug, PartialEq, Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    pub(crate) struct FetchNode {
+        /// The name of the service or subgraph that the fetch is querying.
+        service_name: String,
 
-impl FetchNode {
-    async fn fetch_node<'a>(
-        &'a self,
-        response: Arc<Mutex<Response>>,
-        current_dir: &'a Path,
-        request: Arc<Request>,
-        service_registry: Arc<dyn ServiceRegistry>,
-        schema: Arc<Schema>,
-    ) -> Result<(), FetchError> {
-        let FetchNode {
-            variable_usages,
-            requires,
-            operation,
-            service_name,
-        } = self;
-
-        let query_span = tracing::info_span!("subfetch", service = service_name.as_str());
-
-        if let Some(requires) = requires {
-            // We already checked that the service exists during planning
-            let fetcher = service_registry.get(service_name).unwrap();
-
-            let mut variables = Object::with_capacity(1 + variable_usages.len());
-            variables.extend(variable_usages.iter().filter_map(|key| {
-                request
-                    .variables
-                    .get(key)
-                    .map(|value| (key.clone(), value.clone()))
-            }));
-
-            {
-                let response = response.lock().await;
-                tracing::trace!(
-                    "Creating representations at path '{}' for selections={:?} using data={}",
-                    current_dir,
-                    requires,
-                    serde_json::to_string(&response.data).unwrap(),
+        /// The data that is required for the subgraph fetch.
+        #[serde(skip_serializing_if = "Vec::is_empty")]
+        #[serde(default)]
+        requires: Vec<Selection>,
+
+        /// The variables that are used for the subgraph fetch.
+        variable_usages: Vec<String>,
+
+        /// The GraphQL subquery that is used for the fetch.
+        operation: String,
+    }
+
+    struct Variables {
+        variables: Object,
+        paths: Vec<Path>,
+    }
+
+    impl Variables {
+        fn new(
+            requires: &[Selection],
+            variable_usages: &[String],
+            data: &Value,
+            current_dir: &Path,
+            request: &Request,
+            schema: &Schema,
+        ) -> Result<Variables, FetchError> {
+            if !requires.is_empty() {
+                let mut variables = Object::with_capacity(1 + variable_usages.len());
+                variables.extend(variable_usages.iter().filter_map(|key| {
+                    request
+                        .variables
+                        .get(key)
+                        .map(|value| (key.clone(), value.clone()))
+                }));
+
+                let mut paths = Vec::new();
+                let mut values = Vec::new();
```

Comment:
:grey_question: Do we know the maximum number of values that could go into those vec? Because then we can allocate once for the maximum capacity. It's probably a bit of wasted space but will reduce the number of allocations and we might gain some perf improvement

Reply:
probably the visitor actually knows :thinking: too complicated to implement for now, we would need to go back to an iterator also I believe so never mind. Unless we see a big perf hit here I don't see the value

Reply:
yup I think we can explore the iterator approach later (spoiler alert: I tried, it's annoying because of the `PathElement` enum different behaviours, and I really wish generators were in stable)